spec/fabricators/workshop_fabricator.rb

```diff
@@ -1,86 +1,51 @@
+              host: true)
   end
-end
 
-Fabricator(:virtual_workshop, class_name: :workshop) do
-  date_and_time Time.zone.now + 2.days
-  ends_at Time.zone.now + 2.days + 2.hours
-  title Faker::Lorem.sentence
-  description Faker::Lorem.sentence
-  chapter
-  virtual true
-  student_spaces 10
-  coach_spaces 10
-  slack_channel 'a-channel'
-  slack_channel_link 'https://codebar.slack.link'
-  after_build do |workshop|
-    Fabricate(:workshop_sponsor, workshop: workshop, sponsor: Fabricate(:sponsor), host: false)
-  end
+  transient :student_count
+  transient :coach_count
 end
 
 Fabricator(:workshop_no_sponsor, class_name: :workshop) do
   date_and_time Time.zone.now + 2.days
   ends_at Time.zone.now + 2.days + 2.hours
-  title Faker::Lorem.sentence
-  description Faker::Lorem.sentence
   chapter
 end
 
-Fabricator(:workshop_auto_rsvp_in_past, class_name: :workshop) do
-  date_and_time Time.zone.now + 2.days
-  ends_at Time.zone.now + 2.days + 2.hours
-  title Faker::Lorem.sentence
-  description Faker::Lorem.sentence
-  chapter
+Fabricator(:workshop_auto_rsvp_in_past, from: :workshop) do
   rsvp_opens_at Time.zone.now - 1.day
   invitable false
-  after_build do |workshop|
-    Fabricate(:workshop_sponsor, workshop: workshop, sponsor: Fabricate(:sponsor), host: true)
-  end
 end
 
-Fabricator(:workshop_auto_rsvp_in_future, class_name: :workshop) do
-  date_and_time Time.zone.now + 2.days
-  ends_at Time.zone.now + 2.days + 2.hours
-  title Faker::Lorem.sentence
-  description Faker::Lorem.sentence
-  chapter
+Fabricator(:workshop_auto_rsvp_in_future, from: :workshop) do
   rsvp_opens_at Time.zone.now + 1.day
   invitable false
-  after_build do |workshop|
-    Fabricate(:workshop_sponsor, workshop: workshop, sponsor: Fabricate(:sponsor), host: true)
-  end
 end
 
-Fabricator(:workshop_no_spots, class_name: :workshop) do
-  date_and_time Time.zone.now + 2.days
-  ends_at Time.zone.now + 2.days + 2.hours
-  title Faker::Lorem.sentence
-  description Faker::Lorem.sentence
-  chapter
-  after_build do |workshop|
-    Fabricate(:workshop_sponsor, workshop: workshop, sponsor: Fabricate(:sponsor, seats: 0), host: true)
-  end
+Fabricator(:past_workshop, from: :workshop) do
+  date_and_time 3.months.ago
+  ends_at 3.months.ago + 2.hours
 end
 
-Fabricator(:workshop_no_coaches, class_name: :workshop) do
+Fabricator(:virtual_workshop, class_name: :workshop) do
   date_and_time Time.zone.now + 2.days
-  title Faker::Lorem.sentence
-  description Faker::Lorem.sentence
+  ends_at Time.zone.now + 2.days + 2.hours
   chapter
+  virtual true
+  student_spaces 10
+  coach_spaces 10
+  slack_channel 'a-channel'
+  slack_channel_link 'https://codebar.slack.link'
   after_build do |workshop|
-    Fabricate(:workshop_sponsor, workshop: workshop, sponsor: Fabricate(:sponsor, number_of_coaches: 0), host: true)
+    Fabricate(:workshop_sponsor, workshop: workshop, sponsor: Fabricate(:sponsor), host: false)
```

Comment:
Since this is being refactored, as a virtual workshop does not have to have any sponsors, can we competely remove the after_build callback so that by default tests use the more common virtual workshop scenario of not having any sposors.

Reply:
I introduced a virtual_workshop_sponsored for a couple of tests that needed it.
On the tests that needed change I had the choice of running all the tests as virtual_workshop_sponsored or creating an extra context around the tests that needed the sponsor and making only them virtual_workshop_sponsored. I choose the former to avoid additional nesting which can get confusing for readers of the code - however this is an opinion and I'm happy to do either.